POSIX GNU tar read support

jopadan opened the pull request:
- supporting tared csm.bin of Chasm: The Rift Remastered at https://store.steampowered.com/app/2061230/Chasm_The_Rift/
- See @Panzerschrek [Chasm-Reverse pull request](https://github.com/Panzerschrek/Chasm-Reverse/pull/58) for a working implementation

icculus commented:
This can't be merged, as it contains code licensed under the GPL. Sorry.

sezero commented on src/physfs_tar.h:
```diff
@@ -139,10 +139,10 @@ static bool TAR_encodeOctal(char* data, size_t size, PHYSFS_uint64 i) {
-		return POSIX_FORMAT;
-	return DEFAULT_FORMAT;
+		return (int)POSIX_FORMAT;
+	return (int)DEFAULT_FORMAT;
```
Here is the os/2 (watcom) fix:

```diff
diff --git a/src/physfs_tar.h b/src/physfs_tar.h
index 30b920a..34f1d57 100644
--- a/src/physfs_tar.h
+++ b/src/physfs_tar.h
@@ -7,6 +7,7 @@
 #include <stdbool.h>
 #include <string.h>
 #include <ctype.h>
+#include <sys/stat.h>
 #include <fcntl.h>
 #include <time.h>
 
@@ -139,10 +140,10 @@ static bool TAR_encodeOctal(char* data, size_t size, PHYSFS_uint64 i) {
 
 static int TAR_magic(union block* block) {
 	if(strcmp(block->header.magic, OLDGNU_MAGIC) == 0 && strncmp(block->header.version, TVERSION, 2) == 0)
-		return (int)OLDGNU_FORMAT;
+		return OLDGNU_FORMAT;
 	if(strncmp(block->header.magic, TMAGIC, TMAGLEN - 1) == 0)
-		return (int)POSIX_FORMAT;
-	return (int)DEFAULT_FORMAT;
+		return POSIX_FORMAT;
+	return DEFAULT_FORMAT;
 }
 
 static size_t TAR_fileSize(union block* block) {
diff --git a/src/physfs_archiver_tar.c b/src/physfs_archiver_tar.c
index 732d22a..3b525eb 100644
--- a/src/physfs_archiver_tar.c
+++ b/src/physfs_archiver_tar.c
@@ -27,8 +27,7 @@ static bool TAR_loadEntries(PHYSFS_Io *io, void *arc)
 		}
 
 		/* verify magic */
-		int format = TAR_magic(&current_block);
-		switch(format)
+		switch(TAR_magic(&current_block))
 		{
 			case POSIX_FORMAT:
 				TAR_posix_block(io, arc, &current_block, &count);
@@ -55,14 +54,15 @@ static void *TAR_openArchive(PHYSFS_Io *io, const char *name,
                               int forWriting, int *claimed)
 {
     void *unpkarc = NULL;
+    union block first;
+    enum archive_format format;
 
     assert(io != NULL);  /* shouldn't ever happen. */
 
     BAIL_IF(forWriting, PHYSFS_ERR_READ_ONLY, NULL);
 
-    union block first;
     BAIL_IF_ERRPASS(!__PHYSFS_readAll(io, first.buffer, BLOCKSIZE), NULL);
-    enum archive_format format = TAR_magic(&first);
+    format = TAR_magic(&first);
     io->seek(io, 0);
     *claimed = format == DEFAULT_FORMAT ? 0 : 1;
 
```

sezero commented:
The mode procedures aren't used: I'm guessing that commenting them out will make windows builds succeed.

sezero commented on src/physfs_tar.h:
```diff
@@ -0,0 +1,299 @@
+#include <time.h>
+
+#ifdef __GNUC__
+#if defined(__MINGW32__) || defined(__MINGW64__)
```
MinGW always defines `_WIN32`, so this block shouldn't be needed. `io.h` alone should be sufficient.

However, if this is here only to support the mode functions: see below..

sezero commented on src/physfs_tar.h:
```diff
@@ -0,0 +1,299 @@
+    return sum;
+}
+
+static bool TAR_encodeOctal(char* data, size_t size, PHYSFS_uint64 i) {
```
This isn't actually used: Only referenced by the mode functions which themselves aren't used. can be commented out (also see below)

sezero commented on src/physfs_tar.h:
```diff
@@ -0,0 +1,299 @@
+	return TAR_decodeOctal(block->header.mtime, 12);
+}
+
+mode_t TAR_mode(union block* block)
```
The mode functions `TAR_mode` and `TAR_set_mode` aren't used at all: can be commented out. (also see above.)

sezero commented on src/physfs_tar.h:
```diff
@@ -0,0 +1,299 @@
+		     | (v &  TGEXEC ? S_IXGRP : 0)
+		     | (v &  TOREAD ? S_IROTH : 0)
+		     | (v & TOWRITE ? S_IWOTH : 0)
+	#if !defined(_WIN32) && !defined(__MINGW32__) && !defined(__MINGW64__)
```
See above about windows identification checks: Only `_WIN32` is enough. But the mode procedures aren't used as I noted above..

sezero commented on src/physfs.c:
```diff
@@ -10,7 +10,7 @@
 #define __PHYSICSFS_INTERNAL__
 #include "physfs_internal.h"
-
+#include <time.h>
```
Is this addition really necessary?

sezero commented:
The following patch makes a clean-up of the whole thing:

```diff
diff --git a/src/physfs_tar.h b/src/physfs_tar.h
index e15bc68..aca7e40 100644
--- a/src/physfs_tar.h
+++ b/src/physfs_tar.h
@@ -1,15 +1,7 @@
 #ifndef _INCLUDE_PHYSFS_TAR_H_
 #define _INCLUDE_PHYSFS_TAR_H_
 
-#include <stdio.h>
-#include <stdlib.h>
-#include <stdint.h>
 #include <stdbool.h>
-#include <string.h>
-#include <ctype.h>
-#include <sys/stat.h>
-#include <fcntl.h>
-#include <time.h>
 
 #ifndef PATH_MAX
 #define PATH_MAX 1024
@@ -132,21 +124,15 @@ static PHYSFS_uint64 TAR_decodeOctal(char* data, size_t size) {
     return sum;
 }
 
-static bool TAR_encodeOctal(char* data, size_t size, PHYSFS_uint64 i) {
-    if(	snprintf(data, size, "%llo", i) < 0 )
-    	return false;
-    return true;
-}
-
 static int TAR_magic(union block* block) {
 	if(strcmp(block->header.magic, OLDGNU_MAGIC) == 0 && strncmp(block->header.version, TVERSION, 2) == 0)
-		return (int)OLDGNU_FORMAT;
+		return OLDGNU_FORMAT;
 	if(strncmp(block->header.magic, TMAGIC, TMAGLEN - 1) == 0)
-		return (int)POSIX_FORMAT;
-	return (int)DEFAULT_FORMAT;
+		return POSIX_FORMAT;
+	return DEFAULT_FORMAT;
 }
 
-static size_t TAR_fileSize(union block* block) {
+static PHYSFS_uint64 TAR_fileSize(union block* block) {
 	return TAR_decodeOctal(block->header.size, sizeof(block->header.size));
 }
 
@@ -169,12 +155,12 @@ static bool TAR_checksum(union block* block) {
 	return (reference_chksum == unsigned_sum || reference_chksum == signed_sum);
 }
 
-time_t TAR_time(union block* block)
+static PHYSFS_uint64 TAR_time(union block* block)
 {
 	return TAR_decodeOctal(block->header.mtime, 12);
 }
 
-bool TAR_posix_block(PHYSFS_Io* io, void* arc, union block* block, PHYSFS_uint64* count)
+static bool TAR_posix_block(PHYSFS_Io* io, void* arc, union block* block, PHYSFS_uint64* count)
 {
 	static bool long_name = false;
 	const PHYSFS_Allocator* allocator = PHYSFS_getAllocator();
```


`long_name` having persistent storage in `TAR_posix_block` is a problem.
Haven't tried to follow the logic much, but if you truly want following
it then I suggest passing it as a reference from its caller, e.g. like
below: (compile-tested only !!)

```diff
diff --git a/src/physfs_archiver_tar.c b/src/physfs_archiver_tar.c
index e9ad4f0..03dcab6 100644
--- a/src/physfs_archiver_tar.c
+++ b/src/physfs_archiver_tar.c
@@ -17,6 +17,7 @@ static bool TAR_loadEntries(PHYSFS_Io *io, void *arc)
 	union block zero_block;
 	union block current_block;
 	PHYSFS_uint64 count = 0;
+	bool long_name = false;
 
 	memset(zero_block.buffer, 0, sizeof(BLOCKSIZE));
 	memset(current_block.buffer, 0, sizeof(BLOCKSIZE));
@@ -31,7 +32,7 @@ static bool TAR_loadEntries(PHYSFS_Io *io, void *arc)
 		switch(TAR_magic(&current_block))
 		{
 			case POSIX_FORMAT:
-				TAR_posix_block(io, arc, &current_block, &count);
+				TAR_posix_block(io, arc, &current_block, &count, &long_name);
 				break;
 			case OLDGNU_FORMAT:
 				break;
diff --git a/src/physfs_tar.h b/src/physfs_tar.h
index aca7e40..63f2974 100644
--- a/src/physfs_tar.h
+++ b/src/physfs_tar.h
@@ -160,9 +160,8 @@ static PHYSFS_uint64 TAR_time(union block* block)
 	return TAR_decodeOctal(block->header.mtime, 12);
 }
 
-static bool TAR_posix_block(PHYSFS_Io* io, void* arc, union block* block, PHYSFS_uint64* count)
+static bool TAR_posix_block(PHYSFS_Io* io, void* arc, union block* block, PHYSFS_uint64* count, bool *long_name)
 {
-	static bool long_name = false;
 	const PHYSFS_Allocator* allocator = PHYSFS_getAllocator();
         char name[PATH_MAX] = { 0 };
 	PHYSFS_sint64 time  = 0;
@@ -190,10 +189,10 @@ static bool TAR_posix_block(PHYSFS_Io* io, void* arc, union block* block, PHYSFS
 	/* add file type entry */
 	if (block->header.typeflag == REGTYPE || block->header.typeflag == 0) {
 		/* support long file names */
-		if(long_name) {
+		if(*long_name) {
 			strcpy(&name[0], block->header.name);
 			BAIL_IF_ERRPASS(!__PHYSFS_readAll(io, block->buffer, BLOCKSIZE), 0);
-			long_name = false;
+			*long_name = false;
 			(*count)++;
 		}
 		size = TAR_fileSize(block);
@@ -214,7 +213,7 @@ static bool TAR_posix_block(PHYSFS_Io* io, void* arc, union block* block, PHYSFS
 	/* long name mode */
 	else if(block->header.typeflag == GNUTYPE_LONGNAME)
 	{
-		long_name = true;
+		*long_name = true;
 	}
 	else
 	{
```

The rest is up to @icculus.

jopadan commented:
Thanks for the changes and hints I've integrated them into the pull request.

sezero commented:
> @sezero: Thanks for the changes and hints I've integrated them into the pull request.

Thanks: all my change requests are addressed. It's in @icculus hands now.

Commit:
POSIX GNU tar read support

- supporting tared csm.bin of Chasm: The Rift Remastered at https://store.steampowered.com/app/2061230/Chasm_The_Rift/
- Thanks to @sezero for OS/2 and Windows fixes

jopadan commented:
> This can't be merged as-is, as it contains code licensed under the GPL. Sorry.
I've modified the code not to contain the original code licensed under the GPL now.
zlib license seems to be compatible with GPL but not the other way around.

Commit:
Merge branch 'icculus:main' into main

lndpj committed:
Merge branch 'main' into main

lndpj committed:
update mail

lndpj committed:
Merge branch 'icculus:main' into main

lndpj committed:
Merge branch 'icculus:main' into main
